libs/vm_monitor/src/runner.rs

```diff
@@ -408,33 +410,53 @@ impl Runner {
                         Err(e) => bail!("failed to receive kill signal: {e}")
                     }
                 }
-                // we need to propagate an upscale request
-                request = self.dispatcher.request_upscale_events.recv(), if self.cgroup.is_some() => {
-                    if request.is_none() {
-                        bail!("failed to listen for upscale event from cgroup")
+
+                // New memory stats from the cgroup, *may* need to request upscaling, if we've
+                // exceeded the threshold
+                result = self.cgroup.as_mut().unwrap().watcher.changed(), if self.cgroup.is_some() => {
```

Comment:
The docs say

> Additionally, each branch may include an optional if precondition. If the precondition returns false, then the branch is disabled. The provided <async expression> is still evaluated but the resulting future is never polled. This capability is useful when using select! within a loop.

https://docs.rs/tokio/latest/tokio/macro.select.html

So, I guess the `.unwrap()` here would panci?

(I haven't used this feature of select! before.)

Reply:
Oh yikes, I hadn't realized that before. And yeah, it generally isn't worth making the cgroups stuff optional, just a holdover from earlier designs.

Unfortunately, this bad `tokio::select!` is already the current behavior, which was implemented in #5441, so I'm inclined to ignore this for now and address it in a follow-up.
(Broadly there's significant refactoring that I'd like to do here [particularly in pursuit of neondatabase/cloud#6701], but haven't had the chance to get to that yet)